src/SignalR/common/Protocols.Json/src/Protocol/JsonHubProtocol.cs

```diff
@@ -0,0 +1,758 @@
+        public bool TryParseMessage(ref ReadOnlySequence<byte> input, IInvocationBinder binder, out HubMessage message)
+        {
+            if (!TextMessageParser.TryParseMessage(ref input, out var payload))
+            {
+                message = null;
+                return false;
+            }
+
+            message = ParseMessage(payload, binder);
+
+            return message != null;
+        }
+
+        /// <inheritdoc />
+        public void WriteMessage(HubMessage message, IBufferWriter<byte> output)
+        {
+            WriteMessageCore(message, output);
+            TextMessageFormatter.WriteRecordSeparator(output);
+        }
+
+        /// <inheritdoc />
+        public ReadOnlyMemory<byte> GetMessageBytes(HubMessage message)
+        {
+            return HubProtocolExtensions.GetMessageBytes(this, message);
+        }
+
+        private HubMessage ParseMessage(ReadOnlySequence<byte> input, IInvocationBinder binder)
+        {
+            try
+            {
+                // We parse using the Utf8JsonReader directly but this has a problem. Some of our properties are dependent on other properties
+                // and since reading the json might be unordered, we need to store the parsed content as JsonDocument to re-parse when true types are known.
+                // if we're lucky and the state we need to directly parse is available, then we'll use it.
+
+                int? type = null;
+                string invocationId = null;
+                string target = null;
+                string error = null;
+                var hasItem = false;
+                object item = null;
+                var hasResult = false;
+                object result = null;
+                var hasArguments = false;
+                object[] arguments = null;
+                string[] streamIds = null;
+                JsonDocument argumentsToken = null;
+                JsonDocument itemsToken = null;
+                JsonDocument resultToken = null;
+                ExceptionDispatchInfo argumentBindingException = null;
+                Dictionary<string, string> headers = null;
+                var completed = false;
+
+                var reader = new Utf8JsonReader(input, isFinalBlock: true, state: default);
+
+                reader.CheckRead();
+
+                // We're always parsing a JSON object
+                reader.EnsureObjectStart();
+
+                do
+                {
+                    switch (reader.TokenType)
+                    {
+                        case JsonTokenType.PropertyName:
+                            if (reader.TextEquals(TypePropertyNameBytes))
+                            {
+                                type = reader.ReadAsInt32(TypePropertyName);
+
+                                if (type == null)
+                                {
+                                    throw new InvalidDataException($"Missing required property '{TypePropertyName}'.");
```

Comment:
I'm confused by this. Why would ReadAsInt32 return null, and why does it mean there is no type property? You've just checked you're in the type property so that can't be true.

Reply:
I looked up ReadAsInt32 https://github.com/aspnet/AspNetCore/blob/9de42d516e6ee9c8eb17700e134f3ac8a7a682ff/src/SignalR/common/Shared/SystemTextJsonExtensions.cs#L86

This message would be more accurate to say that the JSON type is wrong like... https://github.com/aspnet/AspNetCore/blob/9de42d516e6ee9c8eb17700e134f3ac8a7a682ff/src/SignalR/common/Shared/SystemTextJsonExtensions.cs#L97
...instead of saying the property is missing.